feat: support array_insert

SemyonSinchenko opened the pull request:
## Which issue does this PR close?
Related to #1042 

`array_insert: SELECT array_insert(array(1, 2, 3, 4), 5, 5)`

## Rationale for this change
As described in #1042 

## What changes are included in this PR?

- `QueryPlanSerde.scala`: I added an additional case for the array insert;
- `expr.proto`: I added a new message for the `ArrayInsert`;
- `planner.rs`: I added a case for the array_insert;
- `list.rs`:
    * I added a new `ArrayInsert` struct;
    * I implemented `PhysicalExpr`, `Display` and `PartialExpr` for it;
    * The main logic of insertion is in `fn array_insert`


## How are these changes tested?
At the moment I added a simple tests for `fn array_insert` and a test for `QueryPlanSerde`.

SemyonSinchenko commented:
As I was able to realize, `array_insert` does not supported in datafusion. Is the `list.rs` a good place to have an implementation of `ArrayInsert` and `PhysicalExpr` for it?

andygrove commented on spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:
```diff
@@ -2313,4 +2313,21 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
     }
   }
+
+  test("ArrayInsert") {
```
You can add an `assume` call here to skip the test for older Spark versions.

```suggestion
  test("ArrayInsert") {
    assume(isSpark34Plus)
```

andygrove commented:
Thanks, @SemyonSinchenko. I think it's fine to skip the test for Spark 3.3. I plan on reviewing this PR in more detail tomorrow, but it looks good from an initial read.

SemyonSinchenko commented:
This PR is ready for the review. The only failed check failed due to the internal GHA error:
```sh
GitHub Actions has encountered an internal error when running your job.
```

andygrove commented on spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:
```diff
@@ -2337,4 +2337,35 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
+    withTempDir { dir =>
+      val path = new Path(dir.toURI.toString, "test.parquet")
+      makeParquetFileAllTypes(path, dictionaryEnabled = false, 10000)
+      spark.read
+        .parquet(path.toString)
+        .withColumn("arr", array(col("_4"), lit(null), col("_4")))
+        .withColumn("idx", udf((x: Int) => x).apply(col("_4")))
+        .withColumn("arrUnsupportedArgs", expr("array_insert(arr, idx, 1)"))
```
This code appears to construct a DataFrame but does not actually execute a query? Also there are not any assertions to check for expected behavior, so this looks like it needs some more work.

SemyonSinchenko replied:
Done!

SemyonSinchenko commented:
Thanks for comments and suggestions!

This PR is ready for the review again.

What were changed from the last round of the review:
- I added tests for the `array_prepend` (spark 3.5+) and enabled the test for `array_append` (spark 4.0+);
- I added tests for the corner cases:
    * Index is negative;
    * Index is bigger than the array length;
    * Index is negative and it's abs is greater than the array length;
    * Test of the fallback to spark (udf as child);
    * Value to insert is actually `null`;
- I fixed the native part behavior for some of the corner cases;

These tests pointed my attention to the uncovered parts of the code that I fixed and I also added couple of additional tests to the `native` part. I revisited how spark do array_insert and it is more tricky than I realized at the first glance. I added few additional comments to the native part of the implementation and fixed the behavior.

At the moment this PR is tested in multiple ways:
- Basic tests in native part that are useful because easy to debug and very fast to run;
- Basic tests in native part for the so called "legacy mode" in spark;
- Basic tests on the scala side;
- Logical tests on small data for corner cases (negative, positive, long, short, null, etc.) on the scala side;
- Tests in array_prepend and array_append that are calling `array_insert` under the hood (NULLS, different data types, etc.);
- Test of the fallback to the Spark in case when one of children is not supported by Comet;

So, it looks to me, that all the possible cases are covered and the behavior is the same like in spark.

SemyonSinchenko commented:
> I am not sure if this should be done together with this PR but it would some add "free" tests. Spark introduced with 3.5 array_prepend which it implements with array_insert and starting 4.0 it also implements array_append with array_insert. If you want you can copy the array_append tests and replace array_append with array_prepend for Spark 3.5+ and enable the array_append tests for spark 4.0. You can also ignore this comment if you do not agree or if it leads to unrelated errors.

Done!

andygrove commented on native/spark-expr/src/list.rs:
```diff
@@ -413,14 +426,297 @@ impl PartialEq<dyn Any> for GetArrayStructFields {
+            .src_array_expr
+            .evaluate(batch)?
+            .into_array(batch.num_rows())?;
+        let src_element_type = match src_value.data_type() {
+            DataType::List(field) => field.data_type(),
+            DataType::LargeList(field) => field.data_type(),
+            data_type => {
+                return Err(DataFusionError::Internal(format!(
+                    "Unexpected src array type in ArrayInsert: {:?}",
+                    data_type
+                )))
+            }
```
minor nit: this logic for extracting a list type is repeated a few times and could be factored out into a function

SemyonSinchenko replied:
@andygrove Thanks for the suggestion!
I moved a checking of the array type (and the exception logic) to the method:
```rs
    pub fn array_type(&self, data_type: &DataType) -> DataFusionResult<DataType> {
        match data_type {
            DataType::List(field) => Ok(DataType::List(Arc::clone(field))),
            DataType::LargeList(field) => Ok(DataType::LargeList(Arc::clone(field))),
            data_type => {
                return Err(DataFusionError::Internal(format!(
                    "Unexpected src array type in ArrayInsert: {:?}",
                    data_type
                )))
            }
        }
    }
```
It allows at least to avoid returning the same error multiple time. Is it what you suggested? Or should I move this method to a helper function and refactor also `GerArrayStructField` to use such a function?

P.S. Sorry for the stupid question... But can you please explain to me why we always check both `List` and `LargeList`, while Apache Spark only supports `i32` indexes for arrays (max length is `Integer.MAX_VALUE - 15`), which is the case of `List` to my understanding? All the code in the `list.rs` might become a bit simpler if we make it non-generic.

andygrove replied:
That's a good question. I wonder if the existing code for LargeList is actually being tested. It would be interesting to try removing it and see if there are any regressions. It makes sense to only handle List if Spark only supports i32 indexes.

Kimahriman replied:
The difference I think is that a `LargeList` can store more than `Integer.MAX_VALUE` entries in _all rows in a single batch_, so if you have multiple Spark rows all with the max num of rows supported, it wouldn't fit into an Arrow `List` array. That would probably need to be supported elsewhere, but it may be worth keeping the `LargeList` handling around in case that scenario is supported. And other DataFusion expressions might return a `LargeList` even if it doesn't come directly from Spark? Does the native Parquet reader ever use a `LargeList`?

SemyonSinchenko replied:
Thanks for the explanation!
You are right, I will close #1118 then

andygrove reviewed (approved):
LGTM. Thanks @SemyonSinchenko!
